### PR TITLE
[Merged by Bors] - feat (Mathlib.Topology.MetricSpace.ThickenedIndicator): Add convergence of plain old indicators.

### PR DESCRIPTION
**Pull request by kkytola**
Add lemmas `tendsto_indicator_thickening_indicator_closure` and `tendsto_indicator_cthickening_indicator_closure`.

Co-authored-by: @sgouezel and @urkud

---

This was PR [mathlib3#17648](https://github.com/leanprover-community/mathlib/pull/17648). There were reviews by @sgouezel and @urkud, but I temporarily abandoned the PR during the semester and the ongoing port to Mathlib4. This Mathlib4 PR takes into account the review comments from the mathlib3 PR.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by ADedecker** [comment]
My involvement into the filter cult forces me to mention that the "true" underlying result is that, in a regular space, `Set.indicator` tends to `Set.indicator (closure E)` along the filter `(𝓝ˢ E).smallSets ⊓ 𝓟 (𝒫 E)`, and that you can deduce your result using the fact that `Metric.thickening` tends to this filter along `𝓝 0` :grin: That wouldn't actually require *that much* new API, but of course I don't want to hold this PR for that at all!

**Review by ADedecker — APPROVED** [review]
I left one suggestion and two answers to comments by Sébastien in the Lean3 PR, otherwise this looks good to me.

Do you want to wait for Yury and/or Sébastien to comment, or is it fine if I "maintainer merge" this?

File: Mathlib/Topology/MetricSpace/ThickenedIndicator.lean
Comment (kkytola) [reply Mathlib/Topology/MetricSpace/ThickenedIndicator.lean]: Thanks! I will create separate docstrings for the additive versions in the other statements, too.

File: Mathlib/Topology/MetricSpace/ThickenedIndicator.lean
Comment (ADedecker) [inline Mathlib/Topology/MetricSpace/ThickenedIndicator.lean:269]:
```suggestion
    filter_upwards [mem_nhdsWithin_of_mem_nhds obs, self_mem_nhdsWithin]
      with δ x_notin_thE δ_pos
```
Sorry, I missed that on first read. The whole point of `filter_upwards` is that you don't have to use `Eventually.and` or `inter_mem`!

**Comment by ADedecker** [comment]
> I had indeed implemented only those suggested generalizations in the mathlib3 PR which were provable, and left two unprovable ones unproven (and since Sébastien had also used a very different tone in the unprovable ones, I interpreted them more as rhetorical questions for contrast in the first place). 

Since you didn't answer I assumed that you missed these comments, sorry for the misunderstanding! Now at least we have a written explanation we can link to if needed.

> I hope the file also makes it evident that the chosen formulations were deliberate, for example by the different use "as δ tends to zero" vs "as δ>0 tends to zero" in the docstrings.

Yes that should be fine.

File: Mathlib/Topology/MetricSpace/HausdorffDistance.lean
Comment (kkytola) [inline Mathlib/Topology/MetricSpace/HausdorffDistance.lean:915]: Btw, these kinds of terms are a mouthful, and I still often end up writing them. Clearly this is bad for readability --- is there a more stylish way?

(Although, for mathematically obvious lemmas like this and the next, I'm of the opinion that only the readability of the statement is important, while their trivial proofs can be obfuscated/golfed. But I would understand also the desire to write readable terms in the proof... So if there is an easy style improvement, I'll be happy to learn it!)

Comment (ADedecker) [reply Mathlib/Topology/MetricSpace/HausdorffDistance.lean:915]: There isn't really any universal trick. I guess this is slightly better:
```lean
  obtain ⟨ε, ⟨ε_pos, ε_lt⟩⟩ := exists_real_pos_lt_infEdist_of_not_mem_closure h
  filter_upwards [eventually_lt_nhds ε_pos] with δ hδ
  simp only [thickening, mem_setOf_eq, not_lt]
  exact le_trans (ENNReal.ofReal_le_ofReal hδ.le) ε_lt.le
```
Note that the reason the term looks better is that the mathematical proof is very slightly simpler (we're just using that `ofReal` is monotone, so no need for `ε_pos` to appear.

About the `filter_upwards` line: technically this does exactly the same as the current code, except that you end up with `hδ : δ < ε` instead of `hδ : δ ∈ Iio ε`, allowing dot notation. In general I would suggest using the `Eventually` version over the `_ ∈ ℱ` version in `filter_upwards` when both exist, but it's not really important.

Comment (ADedecker) [reply Mathlib/Topology/MetricSpace/HausdorffDistance.lean:915]: Oh and of course you can also modify the `cthickening` version in a similar way.

File: Mathlib/Topology/MetricSpace/ThickenedIndicator.lean
Comment (ADedecker) [inline Mathlib/Topology/MetricSpace/ThickenedIndicator.lean:273]:
```suggestion
      with δ x_notin_thE _
    simp only [x_notin_thE, not_false_eq_true, mulIndicator_of_not_mem, x_mem_closure]
```

File: Mathlib/Topology/MetricSpace/ThickenedIndicator.lean
Comment (ADedecker) [inline Mathlib/Topology/MetricSpace/ThickenedIndicator.lean:288]:
```suggestion
    simp only [hδ, not_false_eq_true, mulIndicator_of_not_mem, x_mem_closure]
```

**Comment by ADedecker** [comment]
Thanks!

maintainer merge

**Comment by sgouezel** [comment]
bors r+
Thanks!

**Pull request closed without merging**
